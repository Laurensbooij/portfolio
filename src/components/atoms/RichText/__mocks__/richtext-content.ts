export const mockContent = [
  {
    nodeType: "heading-1",
    data: {},
    content: [
      {
        nodeType: "text",
        value: "Heading 1 text",
        marks: [],
        data: {},
      },
    ],
  },
  {
    nodeType: "heading-2",
    data: {},
    content: [
      {
        nodeType: "text",
        value: "Heading 2 text",
        marks: [],
        data: {},
      },
    ],
  },
  {
    nodeType: "heading-3",
    data: {},
    content: [
      {
        nodeType: "text",
        value: "Heading 3 text",
        marks: [],
        data: {},
      },
    ],
  },
  {
    nodeType: "heading-4",
    data: {},
    content: [
      {
        nodeType: "text",
        value: "Heading 4 text",
        marks: [],
        data: {},
      },
    ],
  },
  {
    nodeType: "heading-5",
    data: {},
    content: [
      {
        nodeType: "text",
        value: "Heading 5 text",
        marks: [],
        data: {},
      },
    ],
  },
  {
    nodeType: "heading-6",
    data: {},
    content: [
      {
        nodeType: "text",
        value: "Heading 6 text",
        marks: [],
        data: {},
      },
    ],
  },
  {
    nodeType: "paragraph",
    data: {},
    content: [
      {
        nodeType: "text",
        value: "Paragraph text",
        marks: [],
        data: {},
      },
      {
        nodeType: "text",
        value: "with bold text",
        marks: [
          {
            type: "bold",
          },
        ],
        data: {},
      },
      {
        nodeType: "text",
        value: " regular text",
        marks: [],
        data: {},
      },
      {
        nodeType: "text",
        value: " and italic text",
        marks: [
          {
            type: "italic",
          },
        ],
        data: {},
      },
      {
        nodeType: "hyperlink",
        data: {
          uri: "https://Laurensbooij.com",
        },
        content: [
          {
            nodeType: "text",
            value: " and a Link",
            marks: [],
            data: {},
          },
        ],
      },
    ],
  },
  {
    nodeType: "embedded-asset-block",
    data: {
      target: {
        metadata: {
          tags: [],
        },
        sys: {
          space: {
            sys: {
              type: "Link",
              linkType: "Space",
              id: "brxg4mjq8zsh",
            },
          },
          id: "6QAUq2XwWCKEWSUMi0oZDO",
          type: "Asset",
          createdAt: "2022-10-14T12:20:41.640Z",
          updatedAt: "2022-10-14T12:20:41.640Z",
          environment: {
            sys: {
              id: "master",
              type: "Link",
              linkType: "Environment",
            },
          },
          revision: 1,
          locale: "en-US",
        },
        fields: {
          title: "Porfolio-image-This portfolio-image-2",
          description: "",
          file: {
            url: "//images.ctfassets.net/brxg4mjq8zsh/6QAUq2XwWCKEWSUMi0oZDO/66bd12b5d8afaacd35f1d382e179679f/Porfolio-image-This_portfolio-image-2.png",
            details: {
              size: 224409,
              image: {
                width: 3200,
                height: 800,
              },
            },
            fileName: "Porfolio-image-This_portfolio-image-2.png",
            contentType: "image/png",
          },
        },
      },
    },
    content: [],
  },
  {
    nodeType: "unordered-list",
    data: {},
    content: [
      {
        nodeType: "list-item",
        data: {},
        content: [
          {
            nodeType: "paragraph",
            data: {},
            content: [
              {
                nodeType: "text",
                value: "List item 1.",
                marks: [],
                data: {},
              },
            ],
          },
        ],
      },
      {
        nodeType: "list-item",
        data: {},
        content: [
          {
            nodeType: "paragraph",
            data: {},
            content: [
              {
                nodeType: "text",
                value:
                  "List item 2.",
                marks: [],
                data: {},
              },
            ],
          },
        ],
      },
      {
        nodeType: "list-item",
        data: {},
        content: [
          {
            nodeType: "paragraph",
            data: {},
            content: [
              {
                nodeType: "text",
                value: "List item 3.",
                marks: [],
                data: {},
              },
            ],
          },
        ],
      },
    ],
  },
]