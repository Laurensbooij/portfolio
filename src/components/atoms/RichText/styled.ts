import styled from 'styled-components'

import themeStyles from 'utilities/themeStyles'

export const Container = styled.div`
  * {
    &:first-child {
      margin-top: 0;
    }

    &:last-child {
      margin-bottom: 0;
    }

    b {
      color: ${themeStyles.colors.primaryText};
    }
  }
`

