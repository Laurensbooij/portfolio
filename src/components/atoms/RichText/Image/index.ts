import styled from 'styled-components'

import themeStyles from 'utilities/themeStyles'

const Image = styled.img`
  display: block;
  height: 200px;
  width: 100%;
  margin: 32px 0;
  border: 2px solid ${themeStyles.colors.grey3};
  border-radius: 12px;
  object-fit: cover;

  @media (min-width: ${themeStyles.breakpoints.tablet}) {
    height: 320px;
  }

  @media (min-width: ${themeStyles.breakpoints.desktop}) {
    height: 340px;
    margin: 40px 0;
  }
`

export default Image