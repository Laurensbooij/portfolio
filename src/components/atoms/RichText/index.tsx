import React, { FC } from 'react'
import parse from 'html-react-parser'

import { H1, H2, H3, H4, H5, H6, BodyText, UnorderedList } from '../TextElements'

import Image from './Image'

import { Container } from './styled'

interface richTextProps {
  content: any,
}

const RichText: FC<richTextProps> = ({ content }) => {

  const extractText = (content: any) => {  
    
    let stringsArray = content.map((item: any) => {
      if(item.marks.find((mark: any) => mark.type === 'bold')) {
        return `<b>${item.value}</b>`
      }
      if(item.marks.includes('italic')) {
        return `<i>${item.value}</i>`
      }
      else return item.value
    })

    const joinedString = stringsArray.join('')
    const element = parse(joinedString)

    return element
  }

  const generateListItems = (content: any) => {
    const listItems = content.map((item: any, index: number) => <li key={`${index}-${item.nodetype}`}>{extractText(item.content[0].content)}</li>)
    return listItems
  }

  let formattedContent = content.map((node: any, index: number) => {

    switch(node.nodeType) {
      case 'heading-1':
        return <H1 key={`${index}-${node.nodetype}`}>{extractText(node.content)}</H1>
      case 'heading-2':
        return <H2 key={`${index}-${node.nodetype}`}>{extractText(node.content)}</H2>
      case 'heading-3':
        return <H3 key={`${index}-${node.nodetype}`}>{extractText(node.content)}</H3>
      case 'heading-4':
        return <H4 key={`${index}-${node.nodetype}`}>{extractText(node.content)}</H4>
      case 'heading-5':
        return <H5 key={`${index}-${node.nodetype}`}>{extractText(node.content)}</H5>
      case 'heading-6':
        return <H6 key={`${index}-${node.nodetype}`}>{extractText(node.content)}</H6>
      case 'unordered-list':
        const listItems = generateListItems(node.content)
        return (
          <UnorderedList key={`${index}-${node.nodetype}`}>{listItems}</UnorderedList>
        )
      case 'embedded-asset-block':
        return <Image key={`${index}-${node.nodetype}`} src={node.data.target.fields.file.url} alt={node.data.target.fields.title} />
      default:
        return <BodyText key={`${index}-${node.nodetype}`}>{extractText(node.content)}</BodyText>
    }
  })
  
  return (
    <Container>
      {formattedContent}
    </Container>
  )
} 

export default RichText