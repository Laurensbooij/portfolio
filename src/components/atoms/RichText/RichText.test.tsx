import React from "react"
import { render, screen } from "@testing-library/react"
import RichText from "./index"

import { mockContent } from './__mocks__/richtext-content'

describe("Tests if RichText component renders correctly", () => {
  render(<RichText data-testid="richtext-element" content={mockContent} />)
  const richTextElement = screen.queryByTestId('richtext-element')

  it("should render the correct conent", () => {
    expect(richTextElement).toMatchSnapshot()
  })
})