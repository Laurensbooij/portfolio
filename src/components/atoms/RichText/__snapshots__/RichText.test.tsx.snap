// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests if RichText component renders correctly should render the correct conent 1`] = `null`;
